### PR TITLE
Fixed issue #19206: alphabetical order can be bad with utf8 caracter

### PR DESCRIPTION
**Pull request by Shnoulle**
Fixed issue #19208: alphabetical order delete answers
Dev: use Collator for ordering, check if exist
Dev: use collator_asort or asort to keep keys

### CODE REVIEW

**Comment by Shnoulle** [comment]
- About collator : unsure need to fix more local (es-mx to es_MX for example)
- Seems php-intl is not mandatory currently when install. I add it on warning if it's OK.
- Why not using [setlocale](https://www.php.net/manual/en/function.setlocale) : «The locale information is maintained per process, not per thread.»

**Comment by Shnoulle** [comment]
Without intl: do a DB request ?

**Comment by gabrieljenik** [comment]
> 1. We really must avoid usage of [setlocale](https://www.php.net/manual/en/function.setlocale.php)

Naive question: Why we should avoid it? Could you expand?

File: application/models/Question.php [hunk 1]
```diff
@@ -778,23 +778,37 @@ private function sortAnswerOptions($answerOptions, $language = null)
...
             if (empty($language) || !in_array($language, $this->survey->allLanguages)) {
                 $language = $this->survey->language;
             }
+            $locale = $language;
+            $aLanguageData = getLanguageData();
+            if (isset($aLanguageData[$language]['cldr'])) {
+                $locale = $aLanguageData[$language]['cldr'];
+            }
+            // need php_intl
+            if (function_exists('collator_create')) {
```
Comment (gabrieljenik) [inline application/models/Question.php:792]: Can we create a wrapper for the collator? or maybe for the sorting procedures?
Then we wouldn't be doing function_exists

**Comment by Shnoulle** [comment]
> > 1. We really must avoid usage of [setlocale](https://www.php.net/manual/en/function.setlocale.php)
> 
> Naive question: Why we should avoid it? Could you expand? 

>  The locale information is maintained per process, not per thread. 

Then : if you have 2 different language on 2 different thread : it can broke.

And if you have another php app : you can broke it.

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php:792]: > Can we create a wrapper for the collator? or maybe for the sorting procedures? Then we wouldn't be repeating function_exists so much

You mean 2 different functions with 80% same code ? Unsure on the best option here.

Comment (gabrieljenik) [reply application/models/Question.php:792]: I meant a function like this:
```
collator_sort($sorted, $collator_sort_type, $fallback_sort_type) {
            // need php_intl
            if (function_exists('collator_create')) {
                $collator = collator_create($locale);
                collator_asort($collator, $sorted, collator_sort_type);
            } else {
                Yii::log("For better alphabeticall ordering : you need php-intl.", 'warning');
                asort($sorted, $fallback_sort_type);
            }            
}
```

Comment (Shnoulle) [reply application/models/Question.php:792]: Inside a foreach : creating a new Collator for each foreach ?

Comment (Shnoulle) [reply application/models/Question.php:792]: If i move too just use an uasort inside the foreach.

Comment (gabrieljenik) [reply application/models/Question.php:792]: > Inside a foreach : creating a new Collator for each foreach ?

Yep, you are right. There other ways, like creating a class.
`class CollatorHelper` with a sort method.
But maybe is not needed. Leave it up to you.

**Review by gabrieljenik — APPROVED** [review]
Code structure looks OK
I think a helper would be usefull for the future and to spread out good sorting practices.
Still, could be left out of the scope of the fix.

File: application/models/Question.php (hunk 1 above)
Comment (Shnoulle) [reply application/models/Question.php:792]: It's a foreach for scale : maybe we can do it ? Since we have maximum 2 scale ?

**Comment by Shnoulle** [comment]
> I think a helper would be usefull for the future and to spread out good sorting practices.

New class in https://github.com/LimeSurvey/LimeSurvey/tree/master/application/core/utils ?

Sorting.php ?

Still need the control when install.

**Comment by gabrieljenik** [comment]
> > I think a helper would be usefull for the future and to spread out good sorting practices.
> 
> New class in https://github.com/LimeSurvey/LimeSurvey/tree/master/application/core/utils ?
> 
> Sorting.php ?

Maybe in application/helpers/ ?
sortingHelper.php

**Comment by Shnoulle** [comment]
> Maybe in application/helpers/ ?
> sortingHelper.php

In application/helpers/ it's more for list of function. 
I check for inclusion to

**Comment by Shnoulle** [comment]
- [ ] Move to an helper Class
- [ ] Add the warning in installer

**Comment by Shnoulle** [comment]
`LimeSurvey\Helpers\SortHelper::getInstance($language)->asort($sorted, LimeSurvey\Helpers\SortHelper::SORT_STRING);`

File: application/helpers/SortHelper.php [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    private function __construct($language)
+    {
+        self::$language = $language;
+        if (class_exists('Collator')) {
+            $locale = $language;
+            $languageData = getLanguageData();
+            if (isset($languageData[$language]['cldr'])) {
+                $locale = $languageData[$language]['cldr'];
+            }
+            self::$collator = new \Collator($locale);
```
Comment (gabrieljenik) [inline application/helpers/SortHelper.php:49]: Does it wort caching the collator as in the cases this is called inside a for-loop? Not sure if itis a heavy load thing

**Review by gabrieljenik — APPROVED** [review]
Code looks good.
Would be nice to add a php unit.

**Comment by Shnoulle** [comment]
Must add a master before

**Pull request merged**